File: day6/part2.rb
def count(days, number, cache = {})
  cache[[days, number]] ||= begin
    if days <= number
      1
    else
      count(days - number, 7, cache) + count(days - number, 9, cache)
    end
  end
end

numbers = ARGF.each_line.first.chomp.split(",").map(&:to_i)
puts(numbers.sum { |number| count(256, number) })

File: day6/part2.py
from functools import cache

@cache
def count(days, number):
    return (
        1 if days <= number
        else count(days - number, 7) + count(days - number, 9)
    )

import fileinput

first_line = str(next(fileinput.input()).strip())
numbers = [int(i) for i in first_line.split(",")]
print(sum(count(256, number) for number in numbers))
